### DISCUSSION TITLE
mergeRotatedTranslatedPage doubt

### DISCUSSION THREAD

**Discussion by josejachuf** [py-pdf/pypdf#903] [Q&A]
Hello everyone

I can't understand what the process is like when applying page1.mergeRotatedTranslatedPage (page2, ...)

Assuming the two pages of equal size
1) Given page1 portrait (rotated 90 °) and page2 landscape

- Are they considered aligned bottom-left?
- What is the rotation point?
- Once rotated, it is transferred by tx and ty?
- Does all of the above apply to page2?

The following works well:
```
page1.mergeRotatedTranslatedPage (
    page2,
    rotation = 90,
    tx = page1.mediaBox.getWidth () / 2,
    ty = page1.mediaBox.getWidth () / 2,
    expand = True
    )
```

2) 1) Given page1 portrait (rotated 270 °) and page2 landscape

This works fine
```
page1.mergeRotatedTranslatedPage (
    page2,
    rotation = 270,
    tx = page1.mediaBox.getHeight () / 2,
    ty = page1.mediaBox.getHeight () / 2,
    expand = True
    )
```

3) The problem I am having with page1 portrait (rotated 270 °) with cropBox! = MediaBox

```
pag1.mergeRotatedTranslatedPage (
    page2,
    rotation = 270,
    tx = page1.mediaBox.getHeight () / 2 + page1.cropBox [0],
    ty = page1.mediaBox.getHeight () / 2 + page1.cropBox [1],
    expand = True
    )
```

more or less it comes out as I want but not as well as I wish

Could someone please explain this to me?
Thank you

**Comment by MartinThoma** [comment]
> more or less it comes out as I want but not as well as I wish

What do you mean by that?

**Comment by Atlasfreak** [comment]
I too have some problems understanding how excatcly this function works. When I am trying to center a landscape page which I rotated to portrait on a bigger page I am absolutley unable to do it as I have no idea what way I need to translate it. I tried to understand the translation matrix that is applied but I was unable to grasp the concept fully.
Additionally I noticed that when the rotation is positive and not 180 degrees you need to use the page height instead of page width but when the rotation is exactly 180 degrees you need to use page width and height with tx and ty respectively.

**Comment by josejachuf** [comment]
I do not know the origin of the PDF that you use, per do you make a recommendation, use the **gs** command to clean inconsistent info containing the PDF before using py-pdf2

`gs -o output.pdf -sDEVICE=pdfwrite -dPDFSETTINGS=/screen input.pdf`

**Comment by Atlasfreak** [comment]
[Test.pdf](https://github.com/py-pdf/PyPDF2/files/8721767/Test.pdf)
Just rotate it and then try to center it on a page that is bigger than A4 while it is possible it is not obvious as it differs from the way you do it with none rotated pages.
With rotated pages when you change tx or ty the page is moved diagonally and not along their respective axis except when it is rotated 180 degrees.
I think this should at least be documented somewhere so people who do not know how transformation matrices work (like me) know how it works and don't need to try arbitrarily until they find something that works.
Side note I found solution for my initial problem.

**Comment by MartinThoma** [comment]
The rotation with `mergeRotatedTranslatedPage` is deprecated now. Please use the following pattern instead (which should give the same results):

```python
from PyPDF2 import Transformation, PdfReader, PdfWriter

reader = PdfReader("box.pdf")
page = reader.pages[0]

x1, y1, x2, y2 = page.cropbox
assert x1 == 0
assert y1 == 0

op = Transformation().rotate(rotation=180).translate(tx=x2, ty=y2)
page.add_transformation(op)

print(page.cropbox)
print(page.mediabox)

writer = PdfWriter()
writer.add_page(page)
with open("output.pdf", "wb") as fp:
    writer.write(fp)
```

The center point of the coordinates is in the bottom-left corner. That means if you want to flip the image by 180 degrees, the content would go from the upper-right quadrant to the lower-left one.

You can see that those are just simple coordinate transformations. PyPDF2 does not care for this operation where the actual content is. It also doesn't care about the mediabox / cropbox / trimbox. It just applies the transformation to the coordinate system.

Did that clarify the topic?

**Reply by carecavoador** [reply]
Hi Martin. First, I want to say I really appreciate your effort in maintaining this. I really like PyPDF2 and I try to use it whenever I can.

I'm having issues with this `add_transformation` method. I took the time to do a [minimal example](https://github.com/carecavoador/min_example) to demonstrate what is going on. The page objects are being translated and rotated, but the cropbox is not.

**Reply by MartinThoma** [reply]
Wow, thank you so much for putting that much effort into investigating the issue! I'll have a look tomorrow (it's late in my timezone and this day was super busy)

**Reply by jassler** [reply]
What do I do / what does it mean if the x1 and y1 coordinates are not 0? I tried to rotate and attach one and the same page below each other, but for some reason the boxes of the pages do not always seem to start at 0, 0.

```python
merger = PdfWriter()

filename = '01.pdf'
outfile = '01_stitched.pdf'

with open(filename, 'rb') as file:
    merger.append(file)
    merger.append(file)

merger.pages[1].rotate(180)
merger.pages[1].transfer_rotation_to_content()

insertion_point = merger.pages[0].mediabox.height
merger.pages[0].merge_translated_page(merger.pages[1], tx=0, ty=-insertion_point, expand=True)
del merger.pages[1]

with open(outfile, 'wb') as fp:
    merger.write(fp)
```

These are the files I tried it on.

* 01.pdf works;
* 02.pdf is basically 01.pdf after it was "stitched" using the script above, but stitching it again it seems to have taken the dimensions of the old file;
* 03.pdf comes from another source, and it also doesn't seem to have the y1 coordinate set to 0.

[01.pdf](https://github.com/user-attachments/files/18364285/01.pdf)
[02.pdf](https://github.com/user-attachments/files/18364287/02.pdf)
[03.pdf](https://github.com/user-attachments/files/18364288/03.pdf)

**Reply by stefan6419846** [reply]
Please avoid chiming in with new requests on old discussions, but instead open a new discussion and link here. Additionally, please provide some more details, like the versions used, executable example code (including the necessary imports) etc. From your code it is not clear how you determined your x1 and y1 values.

**Comment by MartinThoma** [comment]
I thought about creating two more convenience methods for `Page` objects:

* `trim(box)`: Fit a page box (by default box="trimbox", but you could specify "mediabox" or "cropbox" as well) to the actual content.
* `normalize()`: Move the content to the upper-right quadrant, e.g. making sure that there are no negative coordinates + making sure that the lowest x/y coordinates are actually 0.

Would that be useful? Do you have an opinion on the naming of those functions? Which media-

**Reply by Atlasfreak** [reply]
That does sound useful as it would simplyfy automating PDF resizing tasks without havong some unforseen outcomes.

**Reply by raduconst06** [reply]
Hi. Was the normalize feature added?

**Reply by MartinThoma** [reply]
No, sadly not so far.

**Reply by zain910128** [reply]
Hi
Wanted to check if normalise is due to be added soon.

I am struggling with rotation a lot so it would be very helpful.

Thank you

**Reply by MartinThoma** [reply]
I'm sorry, this is not as easy as I thought it would be. I will not be able to work on this feature this year. But maybe somebody wants to add a PR?

**Reply by zain910128** [reply]
Hey Martin,
Thanks for the letting us know.
Can you also provide details on what challenges you've experienced or you anticipate ?

I hope the details will give others an idea of what is involved in case anybody wants to attempt it.

**Reply by MartinThoma** [reply]
You have to (1) get the position and (2) re-compute the position of every single object.

Even if you go with the simpler solution that you just do it via the bounding boxes, you still have to get every single position. It's just too easy to forget something.

**Reply by pubpub-zz** [reply]
Just a (stupid?) question:
What is the advantage of "normalizing". the relative positions of objects are still valid and the `merge_transformed_page()` and other functions seem to do the job. no ?

**Reply by MartinThoma** [reply]
Not a stupid question at all :-) 

When adding an overlay ( https://pypdf.readthedocs.io/en/latest/user/add-watermark.html ) it is not so easy to say how the final result will look like.

For example, say you want to add the watermark in the bottom right corner. If the original images mediabox is (0, 0) to (page_width, page_height) and the watermarks cropbox is (0, 0) to (watermark_width, watermark_height), then you just need to shift the watermark by +(page_width - watermark_width, page_height - watermark_height).

Things are a bit more complicated with the image is not centered. Still doable, though.

**Comment by Cyberware-tech** [comment]
The "minimal example" from [earlier](https://github.com/py-pdf/pypdf/discussions/903#discussioncomment-3159231) in this thread seems to be a broken link.
Can someone please specify a similar example in a more permanent form so that the issue is clear ?

**Reply by pubpub-zz** [reply]
you should find examples in https://pypdf.readthedocs.io

**Reply by zain910128** [reply]
Hey

I actually meant the "minimal example" that @carecavoador posted in the linked post as a description of the bug. The link to that example is broken.

**Reply by pubpub-zz** [reply]
This example deals with PyPDF2 which is obsolete. the core team will not support any discussion about it
